File: examples/src/ticket-simulator.c
// standard library imports
#include <math.h>
#include <pthread.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

// compiler defs
#define BENCH true

// parameters given to worker threads
struct thread_arg {
	int id;
	int seats_a;
	int* tickets;
	pthread_mutex_t* tickets_mut;
};

// function defs
void* work( void* t_arg);
int rand_20();
int rand_1_4();

// entry-point function
int main( int argc, char *argv[]){
	// error check
	if( argc != 4){
		printf( "error: not exactly 3 args\n");
		exit( -1);}

	// parse args
	size_t threads_n = atoi( argv[1]);
	int seats_n = atoi( argv[2]);
	float overselling = atof( argv[3]);
	int seats_a = seats_n + round( seats_n * overselling / 100.0f);

	// initialize tickets pointer and mutex
	int* tickets = malloc( sizeof( int));
	*tickets = 0;
	pthread_mutex_t* tickets_mut = malloc( sizeof( pthread_mutex_t));
	pthread_mutex_init( tickets_mut, NULL);

	// thread vars
	pthread_t threads[threads_n];
	struct thread_arg thread_args[threads_n];
	pthread_attr_t attr;
	pthread_attr_init( &attr);
	pthread_attr_setdetachstate( &attr, PTHREAD_CREATE_JOINABLE);

	// other setup
	srand( time( NULL));

	int error;
	for( size_t i = 0; i < threads_n; i++){
		// initialize thread args
		thread_args[i].id = i;
		thread_args[i].seats_a = seats_a;
		thread_args[i].tickets = tickets;
		thread_args[i].tickets_mut = tickets_mut;

		// start thread
		error = pthread_create(
			&threads[i], &attr, work, (void*) &thread_args[i]);
		// error check
		if( error)
			printf( "error creating thread[%d]: %d\n", i, error);}
	// cleanup
	pthread_attr_destroy( &attr);

	// join on all threads
	for( size_t i = 0; i < threads_n; i++){
		error = pthread_join( threads[i], NULL);
		if( error)
			printf( "error joining thread[%d]: %d\n", i, error);}

	// print summary
	if( ! BENCH)
		printf(
			"Summary: %d tickets sold out of %d (%d + %.1f%%)\n",
			*tickets, seats_a, seats_n, overselling);

	// cleanup and exit
	pthread_mutex_destroy( tickets_mut);
	free( tickets_mut);
	free( tickets);

	//pthread_exit( NULL);
	exit( 0);
}

// worker thread function
void* work( void* t_arg){
	struct thread_arg* arg = (struct thread_arg*) t_arg;
	bool term = false;
	bool even_id = ( arg->id % 2) == 0;
	int threshold = even_id ? 9 : 6;

	while( true){
		pthread_mutex_lock( arg->tickets_mut);
		int seats_left = ( arg->seats_a - *arg->tickets);
		if( seats_left <= 0){
			pthread_mutex_unlock( arg->tickets_mut);
			break;}

		if( rand_20() < threshold){
			int tickets_sold = rand_1_4();
			if( tickets_sold > seats_left)
				tickets_sold = seats_left;
			*arg->tickets += tickets_sold;
			//printf( "thread[%d]: tickets: %d\n", arg->id, *arg->tickets);}
			if( ! BENCH)
				printf(
"Ticket agent %d: Successful transaction - %d tickets sold (%d total)\n",
					arg->id, tickets_sold, *arg->tickets);}
		else if( ! BENCH){
			printf( "Ticket agent %d: Unsuccessful transaction\n", arg->id);}

		pthread_mutex_unlock( arg->tickets_mut);}
	pthread_exit( NULL);}

// return a random number from [0,20)
int rand_20(){
	return rand() % 20;}

// return a random number from [1,4]
int rand_1_4(){
	return 1 + rand() % 4;}

File: examples/src/bin/ticket-simulator.rs
extern crate rand;

use std::sync::{ Arc, Mutex};
use std::thread;

// constants
const BENCH : bool = true;
const AGENT_SLEEP : u32 = 500;

fn main(){
	let threads_n : usize = 1000;
	let seats_n : u32 = 10000;
	let overselling : f32 = 0.0;
	let seats_a : u32 = seats_n + ( seats_n as f32 * overselling) as u32;
	let tickets = Arc::new( Mutex::new( 0));

	let mut handles = vec![];
	for i in 0..threads_n {
		// clone our mutex pointer!
		let tickets = tickets.clone();
		handles.push( thread::spawn(
			move ||{ work( i, seats_a, tickets);}
		));}
	for handle in handles {
		let _result = handle.join();}

	let tickets = tickets.lock().unwrap();
	if ! BENCH {
		println!(
			"Summary: {} tickets sold out of {} ({} + {:.2}%)",
			*tickets, seats_a, seats_n, overselling);}
}

// worker thread function
fn work( id: usize, seats_a: u32, tickets: Arc<Mutex<u32>>){
	let threshold = if id % 2 == 0 { 9 } else { 6 };

	loop {
		if ! BENCH {
			#[allow( deprecated)]
			thread::sleep_ms( AGENT_SLEEP);}
		// lock mutex!
		let mut tickets = tickets.lock().unwrap();
		let seats_left = seats_a - *tickets;
		if seats_left == 0 { break;}

		if rand_20() < threshold {
			let mut tickets_sold = rand_1_4();
			if tickets_sold > seats_left {
				tickets_sold = seats_left;}
			// sell us some tickets!
			*tickets += tickets_sold;
			//printf( "thread[{}]: tickets: {}", id, tickets);}
			if ! BENCH {
				println!(
					"Ticket agent {}: Successful transaction \
					- {} tickets sold ({} total)",
					id, tickets_sold, *tickets);}}
		else {
			if ! BENCH {
				println!( "Ticket agent {}: Unsuccessful transaction", id);}}}}

// return a random number from [0,20)
fn rand_20() -> u32 {
	rand::random::<u32>() % 20}

// return a random number from [1,4]
fn rand_1_4() -> u32 {
	1 + rand::random::<u32>() % 4}